addons/hr_payroll/models/hr_payroll.py

```diff
@@ -97,6 +97,130 @@ def get_all_structures(self):
+        :param date_to: End date of the interval
+        :return: List of tuples with 4 elements ( Code, (date time from, date time to), duration in hours, duration in working days).
+                The code will be set to 'WORK100' for the working time and the holiday type code for the leaves.
+        """
+
+        def was_on_leave_interval(leaves, date_from, date_to, limit=False):
+            """
+                Retrurn a filtered list of the leaves that are in the given period
+            :param leaves: Tuple of 3 values ( date from, date to, type of leave )
+            :param date_from: Date from which the period starts
+            :param date_to: Date on which the period will end
+            :param limit: If set to True, the returned leaves are limited (sliced) between start and end of the period, 
+                        otherwise, the whole period of the leave is returned
+            :return: Tuple of 3 elements ( date from, date to, type of leave )
+            """
+            res = []
+            for leave in leaves:
+                # The cases here are not mutually exclusive, so the order of the test is important
+                #   First we test if the leave is sub-interval of the given interval
+                #   than we test if the leave is overlapping left or right with the interval
+                if leave[0] >= date_from and leave[1] <= date_to:
+                    # If the leave is sub-iterval of the work interval
+                    res.append(leave)
+                elif leave[0] < date_from and leave[1] > date_to:
+                    # If the work interval is sub-interval of the leave
+                    res.append((date_from, date_to, leave[2]))
+                elif leave[0] <= date_to and leave[1] >= date_to:
+                    if limit:
+                        # The case leave[0] > date_from_to is alrady handled with the first if block, so we limit from the start of period
+                        res.append((leave[0], date_to, leave[2]))
+                    else:
+                        res.append(leave)
+                elif leave[0] <= date_from and leave[1] >= date_from:
+                    if limit:
+                        # The case leave[1] < date_to is alrady handled with the first if block, so we limit to the end of period
+                        res.append((date_from, leave[1], leave[2]))
+                    else:
+                        res.append(leave)
+            return res
+
+        def hours_in_interval(interval):
+            """
+                Compute the duration in hours of the intreval
+            :param interval: Interval tuple (date from, date to)
+            :return: Total hours in the interval
+            """
+            return (interval[1] - interval[0]).total_seconds() / 3600.0
+
+        res = []
+        # Read all leaves for the period
+        leave_ids = self.get_resource_leaves(date_from, date_to)
```

Comment:
This and following 3 lines can be replaced in pre_v10 with the _employee_id.iter_leaves_